master/internal/proxy/tls.go

```diff
@@ -0,0 +1,386 @@
+			Organization:  []string{"Determined Master"},
+			Country:       []string{"US"},
+			Province:      []string{""},
+			Locality:      []string{"San Francisco"},
+			StreetAddress: []string{"Golden Gate Bridge"},
+			PostalCode:    []string{"94016"},
+		},
+		IPAddresses: []net.IP{},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(10, 0, 0),
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:    x509.KeyUsageDigitalSignature,
+	}
+
+	signedCert, key, err := genKeyAndSignCert(unsignedCert, masterCACert, false)
+	if err != nil {
+		return err
+	}
+
+	masterKey = key
+	masterCert = signedCert
+	return saveMasterCertAndKey()
+}
+
+// LoadOrGenCA generates a new CA cert and keypair if it does not exist in DB.
+func LoadOrGenCA() error {
+	err := loadCACertAndKey()
+	if err != nil {
+		return err
+	}
+	if masterCACert != nil {
+		return nil
+	}
+
+	logrus.Info("Generating a new CA certificate and key")
+	random, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		return err
+	}
+	caCert := &x509.Certificate{
+		SerialNumber: random,
+		Subject: pkix.Name{
+			Organization:  []string{"Determined Master CA"},
+			Country:       []string{"US"},
+			Province:      []string{""},
+			Locality:      []string{"San Francisco"},
+			StreetAddress: []string{"Golden Gate Bridge"},
+			PostalCode:    []string{"94016"},
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	signedCert, key, err := genKeyAndSignCert(caCert, caCert, true)
+	if err != nil {
+		return err
+	}
+
+	masterCACert = signedCert
+	masterCAKey = key
+	masterCAKeyBytes = x509.MarshalPKCS1PrivateKey(key)
+
+	err = saveCACertAndKey()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetMasterCACert returns the CA cert.
+func GetMasterCACert() ([]byte, error) {
+	err := LoadOrGenCA()
+	if err != nil {
+		return nil, err
+	}
+
+	certBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: masterCACert.Raw,
+	}
+	return pem.EncodeToMemory(certBlock), nil
+}
+
+// GetMasterKeyAndCert returns the key and cert, signed by CA, that Master uses.
+func GetMasterKeyAndCert() (keyPem []byte, certPem []byte, err error) {
+	err = loadOrGenSignedMasterCert()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	keyBlock := &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(masterKey),
+	}
+	certBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: masterCert.Raw,
+	}
+
+	return pem.EncodeToMemory(keyBlock), pem.EncodeToMemory(certBlock), nil
+}
+
+// GenSignedCert generates a key and cert pair, signed by the master CA cert.
+func GenSignedCert() (keyPem []byte, certPem []byte, err error) {
```

Comment:
Same Getter idiom comment as above.

Reply:
Even though it generates a brand new cert, is that still a getter?

Reply:
Oh, sorry, this is `Gen...` not `Get...`.